dhall/src/Dhall.hs

```diff
@@ -976,10 +988,11 @@ instance (Constructor c, GenericInterpret (f :+: g), GenericInterpret h) => Gene
 
         name = constructorModifier (Data.Text.pack (conName n))
 
-        extract u@(UnionLit name' e _)
-            | name == name' = fmap (R1 . M1) (extractR e)
-            | otherwise     = fmap  L1       (extractL u)
-        extract _ = Nothing
+        extract u = do
+          (name', e, _) <- extractUnionConstructor u
+          if
+            | name == name' -> fmap (R1 . M1) (extractR e)
+            | otherwise     -> fmap  L1       (extractL u)
```

Comment:
This instance's implementation is actually not covered by the new tests - GHC seems to prefer to recurse to the right and hence the following instance gets used, though I didn't try very hard to convince it not to.